specs/agents/transport.md

```diff
@@ -2,6 +2,19 @@
 
 Agents send data to the APM Server as JSON (application/json) or ND-JSON (application/x-ndjson) over HTTP. We describe here various details to guide transport implementation.
 
+### User-Agent
+
+In order to help debugging and gathering usage statistics, agents should use one following values for the `User-Agent` HTTP header:
+
+- Header value should start with `elasticapm-${language}/${agent.version}`.
+- If both `service.name` and `service.version` are set, append ` ${service.name}/${service.version}`
+- If only `service.name` is set, append `${service.name}`
+
+Examples:
+- `elasticapm-java/v1.25.0`
+- `elasticapm-ruby/4.4.0 myservice`
+- `elasticapm-python/6.4.0 myservice/v42.7`
+
```

Comment:
Currently the node.js APM agent sends a user-agent looking like:

```
user-agent: elasticapm-node/3.21.1 elastic-apm-http-client/10.0.0 node/12.0.0
```

- The "elastic-apm-http-client/$version" part is an internal library that I could see dropping because it isn't very useful.
- The "node/$version" component seems useful. However, I suppose it is redundant given this in the metadata object: `{"metadata":{"service":{"runtime":{"name":"node","version":"12.0.0"}...`


Is it the intent of this spec to limit the user-agent to *only* these two components?

Reply:
Yes, the _current_ intent of the spec is to limit to only these two components (agent platform + version and service name + version). The way user-agent is intended to be used used here is to provide very simple debugging help, for example to investigate HTTP proxy/load-balancer logs. One example would be to to investigate which agent versions are deployed, or where does the requests come from at high level with the service name. A similar example is described in https://github.com/elastic/apm/issues/509.

However, if there are other attributes that would be relevant to provide here, we could definitely think about adding them also. We probably do not want to provide a generic way to inject variables like we do with transaction/span IDs in logs though.